### PR TITLE
[Manta] Refactor

### PR DESCRIPTION
**Pull request by zhenfeizhang**
* moving out the crypto algorithms to manta-crypto
* moving out the asset definitions to pallet-manta-pay
* better circuit abstraction
* [Important] __better extrinsic interfaces__: this breaks some of current API

### CODE REVIEW

**Comment by zhenfeizhang** [comment]
Will need to add more comments to improve the readability of the code

File: Cargo.toml [hunk 1]
```diff
@@ -38,18 +37,19 @@ ark-serialize = { version = "0.2.0", default-features = false, features = [ "der
...
-
-
 aes = { version = "0.7.0", default-features = false }
 x25519-dalek = { version = "1.1.0", default-features = false, features = [ "u64_backend" ] }
 generic-array = { version = "0.14.0", default-features = false }
 blake2 = { version = "0.9.1", default-features = false }
 
+# Manta Dependencies
+manta-crypto = { branch = "master", git = "https://github.com/Manta-Network/manta-crypto" }
+pallet-manta-asset = { branch = "master", git = "https://github.com/Manta-Network/pallet-manta-asset" }
```
Comment (ghzlatarev) [inline Cargo.toml:44]: Nice.

File: src/payload/mod.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+///     - hash_param: hash parameters.
+///     - pk: proving key of the Groth16 proving system.
+///     - sender_1: meta data for the sender's first coin.
+///     - sender_2: meta data for the second's first coin.
+///     - receiver: a __PROCESSED__ receiver.
+///     - reclaimed_value: the number of reclaimed assets.
+///     - rng: a random number generator.
+/// Outputs:
+///     - a data struct, once serialized, can be passed to the
+///       reclaim intrinsic.
```
Comment (ghzlatarev) [inline src/payload/mod.rs]: reclaim extrinsic*
I think all uses of intrinsic might need to be extrinsic.

Comment (zhenfeizhang) [reply src/payload/mod.rs]: thanks. Fixed/

File: src/zkp/gadget.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+	// step 5. compare the outputs
+	output_var.enforce_equal(&actual_out_var).unwrap();
+}
+
+// =============================
+// circuit for the following statements
+// 1. merkle_path(cm, path, root) == true
+// =============================
+pub(crate) fn merkle_membership_circuit_proof(
```
Comment (stechu) [inline src/zkp/gadget.rs:217]: should this be part of `manta-crypto`?

File: src/zkp/keys.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystem};
+use ark_serialize::CanonicalSerialize;
+use ark_std::rand::{RngCore, SeedableRng};
+use hkdf::Hkdf;
+use manta_crypto::*;
+use pallet_manta_asset::*;
+use rand_chacha::ChaCha20Rng;
+use sha2::Sha512Trunc256;
+use std::{fs::File, io::prelude::*};
+pub fn write_zkp_keys() {
```
Comment (stechu) [inline src/zkp/keys.rs:34]: inline comments/doc of the function?

File: src/zkp/keys.rs [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+	// println!("transfer circuit pk length: {}", transfer_pk_bytes.len());
+
+	let mut reclaim_pk_bytes =
+		manta_reclaim_zkp_key_gen(&hash_param_seed, &commit_param_seed, &rng_seed);
+	let mut file = File::create("reclaim_pk.bin").unwrap();
+	file.write_all(reclaim_pk_bytes.as_mut()).unwrap();
+	// println!("reclaim circuit pk length: {}", reclaim_pk_bytes.len());
+}
+
+fn manta_transfer_zkp_key_gen(
```
Comment (stechu) [inline src/zkp/keys.rs:61]: inline comments/doc of the function?

File: src/zkp/keys.rs [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+	let transfer_vk = &pk.vk;
+	transfer_vk.serialize_uncompressed(&mut vk_buf).unwrap();
+	println!("pk_uncompressed len {}", transfer_pk_bytes.len());
+	println!("vk: {:?}", vk_buf);
+
+	pk.serialize_uncompressed(&mut transfer_pk_bytes).unwrap();
+	transfer_pk_bytes
+}
+
+fn manta_reclaim_zkp_key_gen(
```
Comment (stechu) [inline src/zkp/keys.rs:139]: inline comments/doc of the function?

**Review by stechu — CHANGES_REQUESTED** [review]

File: src/zkp/gadget.rs (hunk 3 above)
Comment (zhenfeizhang) [reply src/zkp/gadget.rs:217]: no. its manta-pay specific circuit. it will create a cyclic dependency if moved.

File: src/zkp/keys.rs (hunk 4 above)
Comment (zhenfeizhang) [reply src/zkp/keys.rs:34]: fixed

File: src/zkp/keys.rs (hunk 6 above)
Comment (zhenfeizhang) [reply src/zkp/keys.rs:139]: fixed

File: src/zkp/keys.rs (hunk 5 above)
Comment (zhenfeizhang) [reply src/zkp/keys.rs:61]: fixed

File: src/test/frame.rs [hunk 7]
```diff
@@ -15,19 +15,20 @@
...
+use pallet_manta_asset::*;
 use rand_chacha::ChaCha20Rng;
 use sp_core::H256;
 use sp_runtime::{
 	testing::Header,
 	traits::{BlakeTwo256, IdentityLookup},
 };
 use std::{boxed::Box, fs::File, io::prelude::*, string::String};
-use x25519_dalek::{PublicKey, StaticSecret};
+// use x25519_dalek::{PublicKey, StaticSecret};
```
Comment (ghzlatarev) [inline src/test/frame.rs]: Forgotten commented out code.

**Comment by zhenfeizhang** [comment]
let's get #40 merged before merging this one

**Review by ghzlatarev — APPROVED** [review]

**Review by stechu — APPROVED** [review]

**Pull request merged**